src/cryptography/hazmat/backends/openssl/x509.py

```diff
@@ -166,6 +172,40 @@ def public_bytes(self, encoding: serialization.Encoding) -> bytes:
         self._backend.openssl_assert(res == 1)
         return self._backend._read_mem_bio(bio)
 
+    def is_issued_by(self, issuer_candidate):
+        if issuer_candidate.subject != self.issuer:
+            raise x509.base.InvalidIssuer(
+                expected=self.issuer, received=issuer_candidate.subject)
```

Comment:
I don't think that the check is sufficient. Other PKIX are typically more rigorous and check KU, SKID, and AKID. NSS seems to be more forgiving than OpenSSL. Ryan mentioned recently that macOS is even more strict and Windows even prioritizes AKID/SKID matching over issuer/subject matching. Even something trivial like "is cert issued by another cert" is full of landmines...

IMHO you should also check for presence of AKID, SKID, and KU, then verify properties like akid.keyid == issuer.skid, akid.serial == issuer.serial, akid.dirname == issuer.issuer, issuer.ku has cert signing flag, and possibly some proxy cert shenanigans. An AKID/SKID mismatch automatically implies cert mismatch.

If I were you I would just use ``X509_check_issued()`` followed by ``X509_verify()``. This gives you OpenSSL semantics, which is what most people expect anyway.

Reply:
@tiran I have given in the [comment above](https://github.com/pyca/cryptography/pull/5116#issuecomment-803466987) reasoning why I think that the mentioned checks are in fact *policy* checks, and there is more than one way to accept them. Say, CA certificate has no AKID, but the issued one has SKID. Is this OK?

> If I were you I would just use `X509_check_issued()` followed by `X509_verify()`. This gives you OpenSSL semantics, which is what most people expect anyway.

This would be definitely the way to go if we were building (yet another) OpenSSL wrapper. But if the project's goal is `to be your "cryptographic standard library"` (and it allows, at least in theory, for other cryptographic backends), then IMO providing the "correct" APIs is a higher priority than making them mimicking OpenSSL. Of course it's up to project maintainers to decide on what is "correct".

Reply:
The x509 layer is outside the hazmat area. Any API outside the hazmat layer should be safe, sane, and free of pitfalls. Your implementation of ``is_issued_by()`` would be a pitfall for me -- and I know a lot about X.509 and PKIX. Even my proposal ``X509_check_issued()`` + ``X509_verify()`` has pitfalls because it doesn't validate EKU and BC extensions.

IMHO it would be better to add low-level functions for X.509 trust chain processing to the hazmat layer.